[TRRFFDA-482] [TRRFFDA-533] Simplify setting up registration for a registry

ScriptSmith opened the pull request.

ScriptSmith committed:
Change default email from address

ScriptSmith committed:
Replace default email value with simple address

ScriptSmith committed:
Use subject as email notification identifier

ScriptSmith committed:
Move RegistryAdmin actions to member methods

ScriptSmith committed:
Simple registration setup action with template, form, and view

ScriptSmith committed:
Interactive registration configuration view

ScriptSmith committed:
Check if cache table has been created

ScriptSmith committed:
Fix default registration notification template formatting

ScriptSmith committed:
Partitioned registration sections

ScriptSmith committed:
Create EmailNotifications

Add from_address field
Added multiple form sections
Use atomic transactions
Expand sections when form returns errors
Identify whether notifications are enabled

ScriptSmith committed:
Use success messages when registration toggled

ScriptSmith committed:
Misc. formatting

ScriptSmith committed:
Additional translation

ScriptSmith commented on rdrf/rdrf/admin.py:
```diff
@@ -221,14 +139,175 @@ def has_delete_permission(self, request, *args, **kwargs):
+    def setup_registration_view(self, request, registry_code):
+        registry = Registry.objects.get(code=registry_code)
+
+        missing_table = self.missing_cache_table()
```
Could be using the walrus operator here, but we're still waiting for flake8 to support it

sztamas replied:
I **love** the walrus!

alintreznai commented on rdrf/rdrf/forms/admin/registry_registration.py:
```diff
@@ -0,0 +1,28 @@
+BootstrapTextAreaInput = Textarea(attrs=_form_attrs)
+
+
+class RegistrationAdminForm(Form):
```
Other registries like MND will need to extend this with other notifications (like ones for primary carer registration). How will this flow work in that case ? From what I see this class and RegistryAdmin will need to be extended and RegistryAdmin.setup_registration_view will need to be overridden along with adding the new templates in settings.py. Is there a way to make this a bit simpler ?

ScriptSmith replied:
I've been operating under the assumption that for complex registries like MND, Angelman, etc. they will want to configure more complex notifications manually. At the moment, this is more of a once-off wizard / interface to quickly see how registration is configured and turn it on and off.

So this process doesn't necessarily need to be extended, as it's more of an abstraction on top of the existing method that helps speed up registration when we are onboarding new registries more frequently.

If there's value in extending it, then I can add a proper view class etc. (perhaps it's worth doing this anyway). You're right that those are the areas I'll need to refactor

alintreznai replied:
I think there is a value in making it extensible. We should consider the base registration flow, notifications and templates defined in TRRF as the basic out of the box solution for new registries which should also be extensible enough to minimize the changes for more complex ones such as MND or Angelman. This will reduce code duplication and serve us better in the long term where more registries will be supported.

ScriptSmith replied:
Ok, in that case I will start making it more extensible and modify the interface such that multiple notifications can be configured.

Should notifications / templates be customisable from this action after they've been set up, or should it point to those admin pages? And if they should be customisable, is the goal to eventually remove those other pages?

alintreznai replied:
Since we already have admin pages to customize notifications/templates I think we should keep this interface simple and only use it for setting up registration. What do you think @sztamas ?

sztamas replied:
Yes, we should re-use the exiting pages as much as possible and just have links to them.

ScriptSmith replied:
We could use dynamic formsets like so:

- https://schinckel.net/2017/02/05/django-dynamic-formsets/
- https://github.com/elo80ka/django-dynamic-formset
- https://stackoverflow.com/a/8097617

However, I think it'd be better to not allow adding multiple notifications at once, unless they're the pre-configured registry-specific notifications.

We could keep all the configuration in the settings file, but as email templates can be quite lengthy, I think it'd be best to make the form & view extensible. And if necessary, we can make adjust the logic if there are more complicated aspects to configure.

ScriptSmith committed:
Move setup_registration_view to its own class

ScriptSmith committed:
Allow setting up multiple registration notifications and templates

ScriptSmith commented:
Allows for multiple notifications to be configured.

WIP so the structure is not final and hasn't been cleaned up yet. Needs to add custom recipients and group recipients to be used by angelman, mnd, etc.

ScriptSmith committed:
Remove features safely

ScriptSmith committed:
Add actions to enable registration, disable registration, and create registration notifications

ScriptSmith committed:
Remove interactive registration setup

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -130,104 +174,142 @@ def export_registry(registry, request):
+            existing_notifications = EmailNotification.objects.filter(
+                registry=registry,
+                description__in=EventType.REGISTRATION_TYPES
+            )
```
This doesn't ignore disabled notifications and also doesn't check for the right registration type.
We should have some business logic code on Registration that checks that the notifications of the right type are enabled for registration to work.
Then that method should be called both from here and from the code that shows the "Registration closed" page if we don't have notifications set up.

ScriptSmith replied:
I missed that it will ignore disabled notifications, but I'm not sure how it doesn't check the right notification type.

I think the redirect to the registration closed page is part of django-registration-redux.

sztamas replied:
Yes, my mistake about the right notification type - it is ok.
We need just account for notifications being disabled.

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -130,104 +174,142 @@ def export_registry(registry, request):
 
-generate_questionnaire_action.short_description = _("Generate Questionnaire")
+        # Check for missing cache table
+        for cache_alias in settings.CACHES:
```
I don't like this code here!

Creating the missing cache table we should check for and do on start-up not having special low-level code for it in unrelated parts of the system.

ScriptSmith replied:
I agree that it's not great to be poking around in undocumented Django internals like this, but I thought this was the least-drastic solution.

Can I add the `createcachetable` command to the docker entrypoint? That way we won't have to preempt the tables not being there.

sztamas replied:
Yes, for now doing it in the docker entrypoint should be fine.

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -130,104 +174,142 @@ def export_registry(registry, request):
+                messages.error(request, _(f"Missing cache table '{cache._table}'. Run django-admin createcachetable"))
 
+        if len(unchanged) > 0:
+            messages.info(request, f"'{', '.join(code for code in unchanged)}' already enabled registration")
```
Any reason we're not adding default notifications for registration ourselves if they don't exist?

ScriptSmith replied:
Do you mean to roll the 'enable registration' and 'create notifications' action into one?

I think they could be, however there are scenarios where the admin will want to enable registration, but not want to create the default notifications.

For example, they want the clinicians to receive the activation email, not the patients. If that's the case, then they will have to delete the new notification as well as create their own one.

I think the warning that notifications need to be created gives enough guidance to the user and doesn't come with any unwanted consequences.

sztamas replied:
OK. But I still don't like it that it is called "Create notifications". It isn't clear to the user what type of notifications we're creating, because there are many types of notifications.
Maybe call it something like "Create default registration notifications"?

Also, while I understand that you might want to avoid the addition of notifications the whole purpose of this feature is to make it easier to enable registration.
Perhaps we could add a button to the notification that calls the "Create default registration notifications" by default for the registries that don't have it.
So you would display the message warning the user the registries 'reg1, reg2' have registration enabled but don't have notifications configured. Would you like to <button>Create default registration notifications</button> or would you like to configure them yourself on the <a>Email Notifications</a> page?

The link would take you to the Email Notifications Admin page, the button would call the admin action that creates the default notifications just as the user selected all the registries that don't have it and selected and execute the Admin action manually.

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -130,104 +174,142 @@ def export_registry(registry, request):
+            registry.add_feature(RegistryFeatures.REGISTRATION)
+            registry.save()
 
+            messages.success(request, f"Registration enabled for '{registry.code}'")
```
For all messages can we use the same display format for the registries as in the table below?
ie. Human-friendly name and code in parens.

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -130,104 +174,142 @@ def export_registry(registry, request):
+                messages.error(request, _(f"Missing cache table '{cache._table}'. Run django-admin createcachetable"))
 
+        if len(unchanged) > 0:
+            messages.info(request, f"'{', '.join(code for code in unchanged)}' already enabled registration")
```
If registration is already enabled the UI currently displays both the "Registration enabled for 'REG_CODE'" and the "REG_CODE" already enabled registration".
In the case, we haven't touched anything we should say just that.

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -130,104 +174,142 @@ def export_registry(registry, request):
+    disable_registration_action.short_description = _("Disable registration")
 
-        return original_urls
+    def create_notifications_action(self, request, registry_models_selected):
```
I am not sure about this admin action.
There are multiple types of notifications, so it isn't really clear what it would do.
I think we shouldn't have a create notifications admin action yet.

We could start by having business logic code to create default registration notifications and we could call it directly from the "Enable registration" action above.

ScriptSmith replied:
See my later comment

sztamas commented on rdrf/rdrf/helpers/notifications.py:
```diff
@@ -0,0 +1,49 @@
+from django.template.loader import get_template
+from django.utils.translation import ugettext as _
+
+
```
I am not sure about all these dataclasses. 
Do we really need them?
They just duplicate the definition we already have in the models.

Can't we just have code that creates the right notifications by creating models?

sztamas replied:
I don't think we need the customisability provided by specifying the NOTIFICATIONS in settings.

The EmailTemplates we could create using datasets in "init_data", similar as we do for users/groups etc.

ScriptSmith replied:
Ok, we can remove the create notifications action and add notifications in the `initial_data` module

sztamas commented on rdrf/rdrf/templates/admin/current_notification_table.html:
```diff
@@ -0,0 +1,54 @@
+{% load i18n admin_static %}
```
I like that you went the extra mile to create a user-friendly view on the existing notifications, but can't we just create filters (filter by registry) on the existing django admin Email Notifications list page and link to those?
That way we wouldn't duplicate existing pages...

ScriptSmith replied:
I wasn't aware of admin filters. I'm happy to use those if I can get them to work.

sztamas reviewed (changes requested):
@ScriptSmith can you please merge next_release into this because it has conflicts?

ScriptSmith committed:
Use registry filters

Note: Filters are not being displayed in the interface, I can't figure
out why. Need to use registry__id__exact=X GET param where X is the pk
of the registry.

ScriptSmith committed:
- Create custom filter for registration notifications
- Redirect to custom filter after registration notifications are created

ScriptSmith committed:
Use standard registry naming scheme

ScriptSmith committed:
Use more descriptive name for notification creation action

ScriptSmith committed:
Don't send 2 messages when registration is already enabled / disabled

ScriptSmith committed:
Add createcachetable command to entrypoint

ScriptSmith committed:
Account for notifications being disabled when registration is enabled

ScriptSmith committed:
Fix translation for action messages

ScriptSmith committed:
Update action message templates

ScriptSmith committed:
Merge branch 'next_release' into nOfOneRegistration

ScriptSmith committed:
Remove checking for missing cache table

ScriptSmith committed:
Use registry's __str__ method

ScriptSmith commented:
I've summarised the changes requested:

- [x] Remove the custom filter in the template and use admin filters
- [ ] Create email templates in `initial_data`
- [ ] Upon enabling registration, the warning popup should link to the create new registration notifications action, and the email notifications admin page
- [x] Rename the 'Create notifications' action
- [x] Use standard registry identifying string
- [x] Don't send 2 messages when registration has already been enabled / disabled
- [x] Add `createcachetable` command to docker entrypoint, and remove existing checking method
- [x] Account for notifications being disabled when checking for existing notifications
- [x] Merge in the new changes in `next_release`

For the 2 remaining changes:

1. Is the idea to create the email templates in initial_data, and then use those as the templates when we create a new notification?
2. Admin actions are a `POST` request, so I think I'd need to use a hidden form or JS to execute the action, and it'd have to use all the correct params and maybe even do csrf stuff. So a link to the action is quite complicated; do you want me to try that out, or ignore using a link for now?

ScriptSmith committed:
Update default notification from address

ScriptSmith committed:
Add default_for_notification field to EmailTemplate

ScriptSmith committed:
Add default_for_notification migration

ScriptSmith committed:
Use templates default templates when creating default notifications

ScriptSmith committed:
Remove default notification dataclasses

sztamas committed:
Merge branch 'next_release' into nOfOneRegistration

ScriptSmith committed:
Merge branch 'next_release' into nOfOneRegistration

sztamas committed:
Merge branch 'nOfOneRegistration' of github.com:eresearchqut/trrf into nOfOneRegistration

sztamas committed:
Making enable registration admin action more readable

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -347,19 +418,43 @@ def groups(self, obj):
+    def queryset(self, request, queryset):
+        if self.value():
+            registry = Registry.objects.get(code=self.value())
+            return queryset.filter(registry=registry, description__in=EventType.REGISTRATION_TYPES)
```
You could do `registry__code=self.value()` directly in the filter and drop line 430.
I also don't think that restricting by `REGISTRATION_TYPES` is OK, as this is a general filter that isn't used just by the registration code, this PR is about.
If you wanted to achieve exactly what you have, ie. display only the registration related notifications, you would have to create a second filter that filter by `EventType`.
Or you could just drop the filtering by `description` at this stage and display all the notifications of the registry. In this case, you would also have to update the messages on the registration admin page "Current registration notifications belonging to the registry:" to "Current notifications belonging to the registry:".

ScriptSmith replied:
> I also don't think that restricting by REGISTRATION_TYPES is OK, as this is a general filter that isn't used just by the registration code, this PR is about.

I'm not sure what you mean by this? The filter is optional, and in the interface it says that it's filtering just registration notifications. It's a general filter, so it's not on by default and can be toggled by the user (when we merge in the filter interface), but it's also used by this PR because that's where the link goes in the `notifications_added.html` template:

```html
<a href="{% url "admin:rdrf_emailnotification_changelist" %}?registry_registration={{ registry.code }}">Link</a>
```

> If you wanted to achieve exactly what you have, ie. display only the registration related notifications, you would have to create a second filter that filter by EventType

In `EmailNotification`, the `description` field is where the type of notification (derived from `EventType`) is stored.

This filter is a custom filter that filters by both registry and registration type (through description). I don't think we could just use two filters, because there isn't an `or` operator in filters that would convey both the registration notification types (and possibly more in the future).

I'm a little confused by what you're asking for here. Perhaps we should discuss in person?

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -347,19 +418,43 @@ def groups(self, obj):
 class EmailNotificationAdmin(admin.ModelAdmin):
     model = EmailNotification
-    list_display = ("description", "registry", "email_from", "recipient", "group_recipient")
+    list_display = ("description", "registry", "email_from", "recipient", "group_recipient", "disabled", "registry")
```
`registry` appears twice in the list.

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -347,19 +418,43 @@ def groups(self, obj):
+        return "-"
+
+
+default_and_lang.short_description = "Default for notification"
```
Can you please put these on the class ie. `EmalTemplateAdmin` instead?

Also, `short_description` isn't translated.

ScriptSmith replied:
I attempted this but you can't refer to `self` in the class variables. I'll try add it as an instance variable in `__init__`.

ScriptSmith replied:
~~Or maybe a lambda :)~~

Nope, it needs the description

sztamas commented on rdrf/rdrf/initial_data/reference_data.py:
```diff
@@ -3,16 +3,30 @@
     AddressType.objects.get_or_create(type="Postal", description="Postal Address")
+
+
+def load_default_notifications():
```
Can you please put this in a separate file (as we will have more to come) and just add it to deps on line 12?
Just as `patient_stage` is done...

sztamas commented on rdrf/rdrf/models/definition/models.py:
```diff
@@ -1353,30 +1353,19 @@ class EmailTemplate(models.Model):
     description = models.TextField()
     subject = models.CharField(max_length=50)
     body = models.TextField()
+    default_for_notification = models.CharField(max_length=100, choices=EMAIL_NOTIFICATIONS, null=True, default=None)
```
This should also have 'blank=True' and you don't need `default=None` which is the default.

ScriptSmith replied:
When I tried using `blank` rather than `null`, there was an not-null constraint exception when creating notifications without a `default_for_notification` value. I'll look into it further.

sztamas commented on rdrf/rdrf/initial_data/reference_data.py:
```diff
@@ -3,16 +3,30 @@
+def load_default_notifications():
+    EmailTemplate.objects.get_or_create(
+        language="en",
+        description="trrf_default_patient_registration_en",
```
This should be a human friendlier text.

sztamas commented on rdrf/rdrf/templates/admin/notifications_disabled.html:
```diff
@@ -0,0 +1,8 @@
+<p>
+  {% trans "Registry " %}'{{ registry.name }}'({{ registry.code }}){% trans " has disabled notifications that need to be enabled for registration to work" %}</p>
+<p>
+  {% trans "Current registration notifications belonging to the registry: " %}
```
Can we change these to "<a>Click here</a> to edit the notifications belonging to this registry"?

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -129,93 +161,133 @@ def export_registry(registry, request):
+            # TODO: Add default notifications for all registration types. Needs different recipients.
+            try:
+                with transaction.atomic():
+                    notification = EmailNotification(
```
You could use `EmailNotification.objects.create(...` here instead and drop the `notification.save()` on the next line.

ScriptSmith replied:
I did that initially but `email_templates` has a many-to-many relationship, so the id has to be created before the templates are specified.

ScriptSmith replied:
```
TypeError: Direct assignment to the forward side of a many-to-many set is prohibited. Use email_templates.set() instead.
```

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -129,93 +161,133 @@ def export_registry(registry, request):
-        "Registry code is readonly after creation"
-        return () if obj is None else ("code",)
+                    notification.email_templates.set(
+                        EmailTemplate.objects.filter(default_for_notification=EventType.NEW_PATIENT)
```
Can we get these once before the `for` loop and also bail out with an error there if there aren't any default notifications set up?

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -129,93 +161,133 @@ def export_registry(registry, request):
+                    }))
+            except IntegrityError as e:
+                logger.error("Failed to create default registration notifications: %s", e)
+                messages.error(request, _("Failed to create default registration notifications"))
```
This error message should include the registry in it, so we know on which registry it failed.

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -129,93 +161,133 @@ def export_registry(registry, request):
-        return original_urls
+    def create_notifications_action(self, request, registry_models_selected):
+        for registry in registry_models_selected:
+            # TODO: Add default notifications for all registration types. Needs different recipients.
```
Shouldn't we check that there are no notifications (of the same type) set up already here and don't add any if so?

ScriptSmith replied:
Well the existing notifications might be for the same event, but are being sent to clinicians rather than patients.

sztamas commented on rdrf/rdrf/admin.py:
```diff
@@ -129,93 +161,133 @@ def export_registry(registry, request):
+                    messages.success(request, render_to_string("admin/notifications_added.html", {
+                        "registry": registry,
+                    }))
+            except IntegrityError as e:
```
What special case are we trying to handle here?
Why are we catching `IntegrityError`s and only `IntegrityError`s?

ScriptSmith replied:
I believe this is a holdover from when it was creating notifications and templates in a single action, but I think it may still be useful here.

As we haven't configured requests to be atomic transactions, I've occasionally run into issues during development where some objects are created during a request, but a failure occurs before the request can complete successfully, and these incomplete objects are left hanging before their relationships to other newly-created objects can be specified.

In cases like this one where we're `save()`ing twice, I think it's a good idea to guarantee that if something goes wrong between saves, things will be rolled back properly.

The `IntegrityError` is intended to catch issues with foreign keys, but we could make it more generic and use `DatabaseError`. Or we can remove the transaction altogether if you'd prefer.

sztamas reviewed (changes requested):
Sorry, somehow lost a few additional comments before, so adding them as a separate review.

ScriptSmith committed:
Move email template fixtures to their own file

ScriptSmith committed:
Remove additional registry field

ScriptSmith committed:
Move default_and_lang to EmailTemplateAdmin class

ScriptSmith committed:
Use friendlier description for default EmailTemplate

ScriptSmith committed:
Change action notification message

ScriptSmith committed:
Error message if default templates have not been created

ScriptSmith committed:
Indicate registry when logging errors

ScriptSmith committed:
Split RegistryRegistrationNotifications into two filters

ScriptSmith committed:
Merge branch 'nOfOneRegistration' of github.com:eresearchqut/trrf into nOfOneRegistration

ScriptSmith committed:
Merge branch 'next_release' into nOfOneRegistration

ScriptSmith committed:
Make default_for_notification blank=True

ScriptSmith committed:
Add blank=True migration

ScriptSmith committed:
Merge branch 'next_release' into nOfOneRegistration

ScriptSmith committed:
Merge migrations
